libtiledbsoma/src/soma/managed_query.cc

```diff
@@ -1244,22 +1405,78 @@ ManagedQuery::_extend_and_evolve_schema_with_details(
         enum_values_in_write.assign(
             (ValueType*)data, (ValueType*)data + num_elems);
     }
+    std::unordered_set<ValueType> unique_values_in_write;
+    for (auto enum_value_in_write : enum_values_in_write) {
+        unique_values_in_write.insert(enum_value_in_write);
+    }
 
-    // Get all the enumeration values in the on-disk TileDB attribute
+    // Check for non-unique values in the input, even before we check the
+    // values in the array schema. See also sc-65078.
+    if (enum_values_in_write.size() != unique_values_in_write.size()) {
+        // std::range_error maps to Python ValueError
+        throw std::range_error(fmt::format(
+            "[extend_enumeration] new values provided for column '{}' must "
+            "be unique within themselves, irrespective of the deduplicate flag",
+            column_name));
+    }
+
+    // Get all the enumeration values in the on-disk TileDB attribute.
     std::vector<ValueType> enum_values_existing = enmr.as_vector<ValueType>();
 
+    // Separate out the values already in the array schema from the
+    // values not already in the array schema.
+    //
+    // One might think it would be simpler to use
+    //
+    //   std::unordered_set<ValueType> existing_enums_set;
+    //
+    // and one would be correct. However, core uses bitwise comparisons, and
+    // floating-point NaNs have the following properties: (1) NaN != NaN, and
+    // (2) there are multiple floating-point bit patterns which are NaN.  It's
+    // simplest to just use the same logic core does, making std::string_view on
+    // our elements.
+    std::unordered_set<std::string_view> existing_enums_set;
+    for (const auto& enum_value_existing : enum_values_existing) {
+        auto sv = std::string_view(
+            static_cast<char*>((char*)&enum_value_existing),
```

Comment:
Why are all ValueTypes being cast to std::string?  (lines 1438-1455)

I understand that you want a "bit value compare" to avoid issues like NaN equality tests, but it seems like casting to a `void*` (a la memcmp) provides clearer intent

Reply:
I hoped my comments were sufficiently clear -- this is explicitly designed to imitate what core is doing.

See also
https://github.com/TileDB-Inc/TileDB/blob/2.27.2/tiledb/sm/array_schema/enumeration.cc#L417-L456